### ISSUE TITLE
[ feature ] Controller blueprints

### ISSUE THREAD

**Issue by JorgeR81** [chrisgoringe/cg-controller#317]
Blueprints would store only the position and placement of Controllers. 

Even with snapping, some configurations may not be very easy to build.
But the same configuration could be useful for different workflows.  

This would be useful for new users. We could include some default blueprints.
This would make it easier to understand the full potential of this extension.

We would have options to:

- Save blueprint ( of the current workflow )
- Load blueprint ( and maybe a list of saved blueprints would show up  )

![template](https://github.com/user-attachments/assets/17e58c23-287b-4579-b332-11534112d7c5)

**Comment by JorgeR81** [comment]
An alternative way to do this,  would be allowing to save empty controllers, in a workflow.

Then we could saved "blueprint" in a workflow.

-----

The disadvantage of this is that "blueprints" could only be used on new workflows.  

Sowe could also have an option to:

- `Load workflow Controllers`

And instead of loading a full new workflow, only its Controller's layout would be loaded, on top of the current workflow.

**Comment by JorgeR81** [comment]
These are the `Load Image` nodes.
One for ControlNet and another for Image2Image.
This way I use different images if I want.
And I can disable them independently on the Controller 


![nodes](https://github.com/user-attachments/assets/718f8f6e-0c7c-4dae-9c38-0469680ab1e5)

---

I have 6 controllers, 1 for each group:

- ControlNet
- Image2Image
- Sampler
- Refiner
- Faced Detailer
- Image Preview

![jpeg](https://github.com/user-attachments/assets/4da36575-b811-4ce9-8d8f-d23acb29929f)

**Comment by JorgeR81** [comment]
I think this has a lot of potential !

Above I talked about 1 `blueprint` for many workflows. But we could also have many `blueprints` for 1 workflow.

Are you familiar with the concept of Blender workspaces?
https://docs.blender.org/manual/en/latest/interface/window_system/workspaces.html

There are default Blender workspaces for common tasks ( e.g. modeling, animation, sculpting, etc. ...) but we can also create custom ones.

`Blueprints` would work like workspaces, but in Sidebar ( instead of Blender tabs ).

Many users have complex workflows that can do multiple things. 
We could use `blueprints` to change the Controller's layout, according to the current task, within the same workflow ( "Inpaint", "Image2Image", etc. ... ) 

---
 
This would require:

- Saving the groups assigned to each Controller, in the `blueprint`
- Saving multiple `blueprints` in one workflow
- Naming a `blueprint` when it's saved
- Having a Sidebar tab with the `blueprint` list ( we could use a floating window, but I think it would look better in the Sidebar )

---

And we could also integrate this with the first approach ( 1 `blueprint` for many workflows ). Like this:

- Save `blueprints` within ComfyUI ( within a single JSON file )
The way it would be easier to share `blueprints` between workflows.
And we could also have some default `blueprints` in there, for new users.
In the Blueprint's Sidebar header we could switch between the 2 lists:
   - `Blueprints` stored in the current workflow
   - `Blueprints` stored in ComfyUI

- When the groups in a `blueprint` don't match the current workflow, they would be discarded, when the `blueprint` is loaded. 

![list](https://github.com/user-attachments/assets/768c1bab-3ea8-4b80-a15d-6a2417688e0e)

**Comment by JorgeR81** [comment]
About the "Blueprint" list name,

I'm fine with any solution you think it's better.

Instead of being in the Sidebar, it could be a dropdown list in the Top bar.
That may be even better, because opening the Sidebar would change the Controller's layout, anyways.

In the old version of the topbar, we had dropdown buttons, and dropdown lists, for the workflows.

![tb3](https://github.com/user-attachments/assets/43e908b7-a608-4fe6-beee-fcb7870c3786)

![tb2](https://github.com/user-attachments/assets/ff245e26-9193-4654-9d0b-4ced1e29b167)

![tb](https://github.com/user-attachments/assets/57c804df-7221-4878-bef1-779112475aaa)

----

Or it could also be in the right click menu and pop up window, like the `Node Templates` list and "manage" window.

![t1](https://github.com/user-attachments/assets/2976c5db-eef2-4b45-8b21-5fe6f6394650)

![tz3](https://github.com/user-attachments/assets/2d95eea3-fde0-44f2-b0e4-c60de068ef90)

**Comment by JorgeR81** [comment]
Or, as alternative, the blueprint list could also be in the right click menu, and pop up window.
This is how the `Node Templates` work, in ComfyUI core.

![t1](https://github.com/user-attachments/assets/2976c5db-eef2-4b45-8b21-5fe6f6394650)

![tz3](https://github.com/user-attachments/assets/2d95eea3-fde0-44f2-b0e4-c60de068ef90)

**Comment by chrisgoringe** [comment]
I was thinking maybe right-click on the controller icon in the topbar

**Comment by JorgeR81** [comment]
> I was thinking maybe right-click on the controller icon in the topbar

Yeah, that could work.

But do you want it to be a dropdown list in the topbar, or a floating window ?

If it's a dropdown list, we could have small arrow button, next to the Controller's icon. Like the old Topbar's `save` icon.

This way it will be more discoverable.

![tb4](https://github.com/user-attachments/assets/563b4f21-a49a-4401-b5aa-9e88a2ce1272)

**Comment by JorgeR81** [comment]
To make this less complex, 

We could only have "_multiple blueprints for 1 workflow_". 

----

Sharing blueprints between workflows ( "_1 blueprint for multiple workflows_" ), may not be worth it, because:
- We would only share the position and placement of the controller windows ( not group tabs ).
- These should be easy enough to recreate, since we would only have to build them once, per workflow.

**Comment by JorgeR81** [comment]
> I was thinking these would be saved as separate files (in the custom node
> directory). I think there’s a pretty simple approach.

Cool, if it's not that much more complex, we could have them both.

**Comment by JorgeR81** [comment]
By the way, I don't mind calling this feature "_workspaces_" ( like Blender ) instead of "_blueprints_".

@chrisgoringe, @LukeG89, What do you think ?

It may be easier to understand the purpose of this feature, if we call it "_workspaces_".

But a possible disadvantage, is that it could to sound too similar to "_workflows_", and users may confuse them both.

**Comment by LukeG89** [comment]
> By the way, I don't mind calling this feature "workspaces" ( like Blender ) instead of "blueprints".

> It may be easier to understand the purpose of this feature, if we call it "workspaces".

Yes, "Workspaces" works better

**Comment by chrisgoringe** [comment]
First cut at this now on the 1.6 branch. Ugly, but seems basically functional.

**Comment by JorgeR81** [comment]
The sizes and positions of the Controller windows seem to be saved correctly. 
Even in very complex workflows.

But the group buttons change location. Some are gone. And some are replaced with an "all" button.
This happens in both simple and complex workflows. 

To test this:

- Exported the workspace.
- Deleted all controllers.
- Imported the workspace again.

---

2 controllers 

**original**


![w01](https://github.com/user-attachments/assets/63a56a6c-e8ab-4fb0-afbb-0f92a3e0b1c9)

---

**imported workspace**

In the bottom windows the "Sampler" button was replaced with a "All" button 

![w02](https://github.com/user-attachments/assets/0c1eb995-136f-4cea-bd59-4e68b185a2f3)


----

7 controllers - full height columns


**original**


![w0](https://github.com/user-attachments/assets/21a50048-5c64-4761-a4f1-8daec4d84df0)

----

**imported workspace**

![w1](https://github.com/user-attachments/assets/5992cf18-052b-470b-a1c7-7173dec2b6bc)


----

18 controllers - stacked columns

**original**

![m02](https://github.com/user-attachments/assets/f9409d40-0134-4337-b742-629af033c3f9)


----

**imported workspace**

![m01](https://github.com/user-attachments/assets/769638dc-a46a-4fc9-acea-298a6ac4838c)

**Comment by JorgeR81** [comment]
About the aesthetics, 

The icons make sense, but they are not very readable. 

Do you like any of these:

![ic0](https://github.com/user-attachments/assets/9b680fcb-5eb5-4878-8b5a-67f2a82fb594)

**Comment by chrisgoringe** [comment]
Expected behaviour is that when you load a workspace all the controllers are empty (or set to All)

That's what I understood the concept to be...

**Comment by JorgeR81** [comment]
> Expected behaviour is that when you load a workspace all the controllers are empty (or set to All)
> 
> That's what I understood the concept to be...

Yes, that's a use case. To share workspaces between workflows. That was my initial idea.
In that case, the group names may be different. 
So we may only need the size and position of the windows. 

---

Another use case would be to have multiple workspaces, for the same workflow.
In that case the group names, would be the same, so it would be useful to store them.

For instance, if I bypass some groups, I can remove their controllers and have a larger controller window for the preview image.
Or if I change the size of the final image ( e.g. portrait to landscape ), I can reconfigure the controller windows, to optimize the space and have a larger preview. 

So, instead of creating new workflows for each controller window variation, I can create workspaces. 
And later on, if I make changes to the main workflow, the workspaces will still be usable.

**Comment by chrisgoringe** [comment]
Moving this to 1.7 so the stuff in 1.6 can move to dev...

**Comment by JorgeR81** [comment]
Here is an example use case.

These are 2 identical workflows. I just changed the image size, and the Controller window configuration. 

I'd like to store both controller configurations, in the same workflow, to quickly change between them.
Now I need to switch between 2 different workflows. 
And each time I need to make changes to the workflow, I need to do it twice. 

---

![p02](https://github.com/user-attachments/assets/3e66e182-9086-41ea-a938-331deff768e7)

![p01c](https://github.com/user-attachments/assets/985bb95b-a0a5-4d16-a3b4-a53766733109)

**Comment by chrisgoringe** [comment]
OK. So the "workspace" should actually be a complete representation of the controllers (position, group, and options). 

That's actually easier, in a way!

I'll aim to get export and import working first, then we can consider some sort of library

**Comment by JorgeR81** [comment]
> I'll aim to get export and import working first, then we can consider some sort of library


Yeah, exporting all Controller data, for a single layout, would give basic functionality. 

----

Ideally, we could have multiple workspaces in the same workflow file ( json / png ).

But, I'm not sure how the GUI should look.
 
Maybe have another Top Bar button, to open a pop up or dropdown with all workspaces. 
We could add, rename, select and delete workspaces from there. 

ComfyUI also uses this icon for ( workflow ) templates. 
And it kind of looks like a "Controller workspace" representation.

![btn](https://github.com/user-attachments/assets/b1d8c140-56dd-4b53-b566-c5719d65e7e8)

----

When we open the "Library", for the first time, we could see that we are in "Workspace 1".

If we add a new one, it could duplicate the workspace that's currently selected.

But when we make new modifications in the Controller, they would be stored separately.
And we could always return to the previous workspace.

The new modifications ( and the new workspace library "entry" ) only need to be "permanently stored" if we `Save` or `Export` the whole workflow via Comfy UI.

**Comment by chrisgoringe** [comment]
Done a bunch more work on this on 1.7. Now includes the list of groups and similar.

There is an issue with node ids not being preserved, so things like node order and widget hiding isn't consistent, so loading a workspace into a new workflow isn't as good as it could be.

**Comment by JorgeR81** [comment]
> There is an issue with node ids not being preserved, so things like node order and widget hiding isn't consistent.

I tried this in a very complex workflow, and I didn't notice these issues:
- the node order was preserved in all controller windows
- the hidden widgets, in each node, did not change

![x0](https://github.com/user-attachments/assets/90fda15c-eba8-4e2d-aebe-34c61481fda2)

**Comment by chrisgoringe** [comment]
The problem occurs when you do the following:

- save a layout for a workflow
- open a new blank sheet
- insert the first workflow
- load the layout

**Comment by JorgeR81** [comment]
I saved 3 workspaces for the same workflow, and I'm switching between them.

The layout seems to be consistent and the node ID doesn't change, when I load a workspace.

Although, the only change I've made was removing some controller windows, and making the preview window larger.

---

![w3](https://github.com/user-attachments/assets/1ac748ed-195c-4d91-b858-4ecd1ca2a89f)

![w2](https://github.com/user-attachments/assets/746be2be-7ea4-48bb-ab8e-b032f9cb9cfb)

![w1](https://github.com/user-attachments/assets/b9164ee8-179e-450c-bcc4-ec25331d548c)

**Comment by JorgeR81** [comment]
> The problem occurs when you do the following:
> 
> - save a layout for a workflow
> - open a new blank sheet
> - insert the first workflow
> - load the layout


Do you mean, using the `insert` option, in the right click menu ?

![ins](https://github.com/user-attachments/assets/175ceb44-bc3a-41b0-b69f-2734642f46f0)

If we want to load the workflow normally, we just need to ( left ) click the workflow name. No need for the context menu.

I think the `insert` option is mostly for adding nodes to an existing workflow ( like a Template would do ).
So, I would expect the node ID to change if I do that.
I don't think that's a problem. 

![temp](https://github.com/user-attachments/assets/416d9324-94d2-4a04-b06a-01ffbf2e16a3)

**Comment by chrisgoringe** [comment]
Yeah, I was just experimenting - taking a workflow (with controller) and
adding it to another workflow doesn’t work very well.

But the other way round does.

It’s not a block, but if I can work out how I want to fix it 😀
